fix(table): applied currentTarget property to event object to retrieve parent element instead of child.

gavin-everett-genesys opened the pull request:
**Description of issue :** 
For sortable columns in the gux-table component, if there is a child element in the sortable th element, and the user clicks on the child element, the guxsortchanged event does not contain the th's attributes for the columnName and sortDirection.

**RCA:**
In the table component the `columnElement` value was being set to `event.target` which was targeting the inner span tag if present instead of the parent `th` element. 

**Resolution:**
As a resolution to this I applied the `currentTarget` property to the event object instead of `target` property.
https://developer.mozilla.org/en-US/docs/Web/API/Event/currentTarget

**View changes :** 
https://apps.inindca.com/common-ui-docs/genesys-webcomponents/feature/COMUI-1017

daragh-king-genesys commented on src/components/beta/gux-table/gux-table.tsx:
```diff
@@ -483,7 +483,7 @@ export class GuxTable {
       if (Object.prototype.hasOwnProperty.call(column.dataset, 'sortable')) {
         column.onclick = (event: MouseEvent) => {
-          const columnElement = event.target as HTMLElement;
+          const columnElement = event.currentTarget as HTMLElement;
```
Would it be possible to add an e2e test to make sure we don't get a regression on this again.

daragh-king-genesys reviewed (commented):
In general this is something I have been thinking about, we are going to need to improve our testing coverage and more importantly depth. I think trying to add tests when we fix bugs like this is a good start/practice that will help us improve.

gavin-everett-genesys commented:
@daragh-king-genesys @katie-bobbe-genesys @MattCheely 
**Updates :**
Fix for firefox : 
On firefox the sorting function never functioned correctly. It emitted the event but never ran the sorting function. The cause of this was on firefox it was running two event listeners at once which affected the function. As a resolution `event. stopImmediatePropagation` was used to stop the remaining listeners from being called.

**E2E Tests :** 
Added two e2e tests. First test checks to see if the `guxsortchanged` event outputs the `first-name` column  with a nested span tag.

Second test checks to see if the `guxsortchanged` event outputs the `first-name` column without a nested span tag.

daragh-king-genesys commented on src/components/beta/gux-table/tests/gux-table.e2e.ts:
```diff
@@ -133,4 +133,117 @@ describe('gux-table-beta', () => {
   });
+
+  it('should sort table if table header nested element is wrapped in a span tag', async () => {
+    const table = `<gux-table-beta id="sortable-table">
```
The `id` attribute in not needed but you should add a `lang` one.

daragh-king-genesys commented on src/components/beta/gux-table/tests/gux-table.e2e.ts:
```diff
@@ -133,4 +133,117 @@ describe('gux-table-beta', () => {
+  </gux-table-beta>`;
+
+    const page = await newE2EPage();
+    await page.setContent(table);
```
I prefer `const page = await newE2EPage({ html: table });` or even better `const page = await newE2EPage({ html });`

daragh-king-genesys commented on src/components/beta/gux-table/tests/gux-table.e2e.ts:
```diff
@@ -133,4 +133,117 @@ describe('gux-table-beta', () => {
+    });
+  });
+
+  it('should sort table if table header nested element is not wrapped in a span tag', async () => {
```
my above comments apply to this test too.

daragh-king-genesys commented on src/components/beta/gux-table/example.html:
```diff
@@ -352,6 +352,7 @@ <h2>Data table with sortable columns</h2>
 
     table.addEventListener('guxsortchanged',
     (event) => {
+      event.stopImmediatePropagation();
```
Why is this needed?

gavin-everett-genesys replied:
The original implementation of the sorting never worked on firefox. After debugging it, it seemed that 2 event listeners were being called at once which prevented the sort from executing on the table. Used `stopImediatePropagation` to prevent other listener from being called.

daragh-king-genesys replied:
I don't think this is a valid solution as it is in the example file not that component.

gavin-everett-genesys replied:
Will look into this.

katie-bobbe-genesys replied:
I agree with Daragh that adding a line to the example file is probably not a valid solution to the bug in Firefox.

gavin-everett-genesys replied:
There are 2 event listeners being created in the example file for `guxsortchanged`. From the callstack in firefox it seems to be executing both so first it will run the `desc`  then `asc` leaving the list the same way it was. In the callstack the issue seems to be primarily with the `onload` in the example file it runs twice at the start and then continues with the rest of the operations. That is why I used stopImediatePropagation to prevent the bubbling but it probably isn't viable solution.

gavin-everett-genesys commented:
@daragh-king-genesys @MattCheely @katie-bobbe-genesys 

Hi all the issue in firefox regarding the table not sorting has now been resolved. I had previously done a fix on another branch for the resizing and sorting issue which turns out fixes the firefox issue also. I rebased this branch to main and the sorting is now working fine on all browsers.

Also on this branch is a fix where if the nested child of the `th` is wrapped in a span tag that it will still sort the table previously it didn't. I also added 2 E2E tests to test this functionality.

katie-bobbe-genesys reviewed (changes requested):
This is a good opportunity to address the sort order accessibility issues as well. 
Here is an example of accessible table sorting: https://dequeuniversity.com/library/aria/table-sortable The accessible table sorting example is not necessarily perfect, but is a good reference point to start from. For reference, here is the ticket that was made to track the table accessibility issues: https://inindca.atlassian.net/browse/COMUI-829

I think that we will need to make a new component for the gux-table. I would suggest experimenting with a few ways to do this. One way would be to make a new `gux-th` component that will render a button within a `th` element when given the 'sortable' attribute. We would continue to support the current API as well, but update our example files to use `gux-th`
A few potential other advantages to adding a new `gux-th` component
1. Adding resizing accessibility will be more straightforward
2. Easier to add tooltips for truncated titles in the future
Please reach out if you have any questions or would like to discuss potential API designs

gavin-everett-genesys commented:
@MattCheely @katie-bobbe-genesys 
This PR is ready to be merged. It consists of a fix for if you have nested span within a `th` element you will still be able to sort the table. A bug was found during the testing of this PR with the original implementation of the sortable table that is you cannot sort the table in firefox. I have created a separate bug that you can view here , https://inindca.atlassian.net/browse/COMUI-1057
